### PR TITLE
Added options to x509 tool

### PR DESCRIPTION
**Pull request by ecdeye**
Description:

 Contains initial implementation for OpenSSL x509 tool, options -in -out, -req, -signkey, -modulus, -days, -dates,
 -checkend, -noout (x509.cc), as well as testing for option functionality and argument requirements (x509_test.cc)

_Files expected to change_

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and the ISC license.

### CODE REVIEW

File: tool-openssl/x509.cc [hunk 1]
```diff
@@ -3,72 +3,251 @@
...
+  }
+  if (!PEM_write_X509(out_file.get(), x509)) {
+    fprintf(stderr, "Error: error writing certificate to '%s'\n", out_path.c_str());
+    ERR_print_errors_fp(stderr);
+    return false;
+  }
+  return true;
+}
+
+bool LoadAndSignCertificate(X509 *x509, const std::string &signkey_path) {
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:41]: Nit: The naming of the function makes it seem like a `Certificate` is being both loaded and signed, but we're actually loading a private key and signing a certificate. Maybe something like `LoadPrivateKeyandSignCertificate`?

File: tool-openssl/x509.cc [hunk 2]
```diff
@@ -3,72 +3,251 @@
...
+    return false;
+  }
+  EVP_PKEY *pkey = PEM_read_PrivateKey(signkey_file.get(), nullptr, nullptr, nullptr);
+  if (!pkey) {
+    fprintf(stderr, "Error: error reading private key from '%s'\n", signkey_path.c_str());
+    ERR_print_errors_fp(stderr);
+    return false;
+  }
+  bssl::UniquePtr<EVP_PKEY> pkey_guard(pkey);
+  if (!X509_sign(x509, pkey, EVP_sha256())) {
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:54]: Just double checking that OpenSSL's default is SHA256 right? This isn't part of this project phase, but it might be customizable with the `-digest` option later on. 
Could we leave a TODO comment above here to help us keep track?

File: tool-openssl/x509.cc [hunk 3]
```diff
@@ -3,72 +3,251 @@
...
+      char not_before[20], not_after[20];
+      ASN1_TIME *nb = X509_getm_notBefore(x509.get());
+      ASN1_TIME *na = X509_getm_notAfter(x509.get());
+      struct tm tm;
+      ASN1_TIME_to_tm(nb, &tm);
+      strftime(not_before, sizeof(not_before), "%Y-%m-%d %H:%M:%S", &tm);
+      ASN1_TIME_to_tm(na, &tm);
+      strftime(not_after, sizeof(not_after), "%Y-%m-%d %H:%M:%S", &tm);
+      printf("notBefore=%s\n", not_before);
+      printf("notAfter=%s\n", not_after);
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:219]: Slight nit: It would be best to let the `printf` and `fprintf` calls be consistent.

File: tool-openssl/x509.cc [hunk 4]
```diff
@@ -3,72 +3,251 @@
...
+    }
+
+    if (checkend) {
+      ASN1_TIME *current_time = ASN1_TIME_set(nullptr, std::time(nullptr));
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:223]: There might not be a C++ destructor for `ASN1_TIME`, but feel free to add one here
https://github.com/aws/aws-lc/blob/4c8e80fea589daeea5f677d9012b37baac871aac/include/openssl/asn1.h#L2082

File: tool-openssl/x509_test.cc [hunk 5]
```diff
@@ -21,17 +24,18 @@
...
 size_t createTempFILEpath(char buffer[PATH_MAX]);
 
+void RemoveFile(const char* path);
+
 X509* CreateAndSignX509Certificate() {
   bssl::UniquePtr<X509> x509(X509_new());
   if (!x509) return nullptr;
 
-  // Set validity period for 1 year
+  // Set validity period for 30 days
```
Comment (samuel40791765) [inline tool-openssl/x509_test.cc:33]: Not an issue, but was curious why the change to 30 days?

File: tool-openssl/x509.cc [hunk 6]
```diff
@@ -3,72 +3,251 @@
...
+    fprintf(stderr, "Error: unable to load private key from '%s'\n", signkey_path.c_str());
+    return false;
+  }
+  EVP_PKEY *pkey = PEM_read_PrivateKey(signkey_file.get(), nullptr, nullptr, nullptr);
+  if (!pkey) {
+    fprintf(stderr, "Error: error reading private key from '%s'\n", signkey_path.c_str());
+    ERR_print_errors_fp(stderr);
+    return false;
+  }
+  bssl::UniquePtr<EVP_PKEY> pkey_guard(pkey);
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:53]: Can we just load `pkey` above with 
```
bssl::UniquePtr<EVP_PKEY> pkey(PEM_read_PrivateKey(signkey_file.get(), nullptr, nullptr, nullptr))
```
?

File: tool-openssl/x509_test.cc [hunk 7]
```diff
@@ -47,59 +51,242 @@ X509* CreateAndSignX509Certificate() {
...
-    std::vector<uint8_t> output_data;
-    ASSERT_TRUE(ReadAll(&output_data, out_file.get()));
+  bssl::UniquePtr<EVP_PKEY> pkey(EVP_PKEY_new());
+  ASSERT_TRUE(pkey);
 
-    // Ensure output data not empty
-    ASSERT_FALSE(output_data.empty());
+  bssl::UniquePtr<RSA> rsa(RSA_new());
+  bssl::UniquePtr<BIGNUM> bn(BN_new());
+  ASSERT_TRUE(bn && BN_set_word(bn.get(), RSA_F4) && RSA_generate_key_ex(rsa.get(), 2048, bn.get(), nullptr));
```
Comment (samuel40791765) [inline tool-openssl/x509_test.cc:137]: Nit: we should check `rsa` here as well

**Review by samuel40791765 — COMMENTED** [review]
Great work and test coverage! Just a few minor comments here and there

File: tool-openssl/x509.cc [hunk 8]
```diff
@@ -3,72 +3,251 @@
...
+    return false;
+  }
+  if (!signkey_path.empty() && (dates || checkend)){
+    fprintf(stderr, "Error: '-signkey' option cannot be used with '-dates' and '-checkend' options\n");
+    return false;
+  }
+  if (days && (dates || checkend)){
+    fprintf(stderr, "Error: '-days' option cannot be used with '-dates' and '-checkend' options\n");
+    return false;
+  }
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:111]: We can consider adding tests for these mutually exclusive options as well

File: tool-openssl/x509_test.cc (hunk 5 above)
Comment (ecdeye) [reply tool-openssl/x509_test.cc:33]: 30 days is the default for OpenSSL

File: tool-openssl/x509.cc (hunk 3 above)
Comment (ecdeye) [reply tool-openssl/x509.cc:219]: updating this and dates output to match OpenSSL output

File: tool-openssl/x509.cc (hunk 4 above)
Comment (ecdeye) [reply tool-openssl/x509.cc:230]: Changed to `bssl::UniquePtr<ASN1_TIME> current_time(ASN1_TIME_set(nullptr, std::time(nullptr)));`

File: tool-openssl/x509.cc [hunk 9]
```diff
@@ -3,72 +3,257 @@
...
+      }
     }
 
-    // Serialize certificate to DER format
-    uint8_t *out_data = nullptr;
-    int len = i2d_X509(x509.get(), &out_data);
-    if (len < 0) {
-        fprintf(stderr, "Failed to serialize X509 certificate.\n");
-        ERR_print_errors_fp(stderr);
+    if (checkend) {
```
[the remark this answers is not in the document]
Comment (justsmth) [reply tool-openssl/x509.cc:230]: Oh.. of course, if negative values are not valid then you could use negative values to indicate that the option was not specified.

File: tool-openssl/x509.cc [hunk 10]
```diff
@@ -3,72 +3,257 @@
...
+  }
+  if (req && (dates || checkend)){
+    fprintf(stderr, "Error: '-req' option cannot be used with '-dates' and '-checkend' options\n");
+    return false;
+  }
+  if (!signkey_path.empty() && (dates || checkend)){
+    fprintf(stderr, "Error: '-signkey' option cannot be used with '-dates' and '-checkend' options\n");
+    return false;
+  }
+  if (days && (dates || checkend)){
```
[the remark this answers is not in the document]
Comment (ecdeye) [reply tool-openssl/x509.cc:112]:
```
>> openssl x509 -in "input_csr.pem" -req -signkey "private_key.pem" -out "output_file3.pem" -days 0
Certificate request self-signature ok
subject=C=AU, ST=Some-State, O=Internet Widgits Pty Ltd

>> openssl x509 -in "output_file3.pem" -dates
notBefore=Jul 12 14:32:33 2024 GMT
notAfter=Jul 12 14:32:33 2024 GMT
```

OpenSSL lets you input -days 0 without error, even though it doesn't make sense to do. Yesterday, I was discussing with Samuel other quirks as a matter of OpenSSL being less restrictive. This tool is already more restrictive than OpenSSL by requiring certain options be used or not used with each other. I can add check that days is not 0 though, but trying to figure out where to balance similar functionality and what is logical to require. Ideally, the customers that are dependent on OpenSSL would not use the tool this way.

[the remark this answers is not in the document]
Comment (ecdeye) [reply tool-openssl/x509.cc:112]: Side note, OpenSSL even lets you input negative days so you can time travel 😃

File: tool-openssl/x509.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (ecdeye) [reply tool-openssl/x509.cc:253]: Ok just fixed the non-negative integer issue for -checkend and non-positive integer issue for -days. Tested them and they seem to work, please let me know thoughts on implementation.

File: tool-openssl/x509.cc [hunk 11]
```diff
@@ -3,72 +3,280 @@
...
     if (!x509) {
-        fprintf(stderr, "Failed to parse X509 certificate from '%s'.\n", in_path.c_str());
-        ERR_print_errors_fp(stderr);
+      fprintf(stderr, "Error: error parsing certificate from '%s'\n", in_path.c_str());
+      ERR_print_errors_fp(stderr);
+      return false;
+    }
+
+    if (dates) {
+      BIO *bio = BIO_new(BIO_s_mem());
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:210]: Nit: We can use `bssl::UniquePtr<BIO> bio(BIO_new(BIO_s_mem()));` here.

File: tool-openssl/x509.cc [hunk 12]
```diff
@@ -3,72 +3,280 @@
...
+    checkend.reset(new unsigned(std::stoul(checkend_str)));
+  }
 
-    // Check for required arguments
-    std::string in_path, out_path;
-    if (!GetString(&in_path, "-in", "", parsed_args)) {
-        fprintf(stderr, "Missing required argument: -in\n");
-        PrintUsage(kArguments);
-        return false;
+  // Check that -checkend argument is valid, int >0
```
Comment (samuel40791765) [inline tool-openssl/x509.cc:129]:
```suggestion
  // Check that -days argument is valid, int > 0
```

File: tool-openssl/x509_test.cc [hunk 13]
```diff
@@ -47,59 +48,336 @@ X509* CreateAndSignX509Certificate() {
...
+  }
+
+  // -noout with -out, -modulues, -dates, -checkend
+  args_list_t args1 = {"-in", in_path, "-noout", "-out", out_path};
+  ASSERT_FALSE(X509Tool(args1));
+  args_list_t args2 = {"-in", in_path, "-noout", "-modulus"};
+  ASSERT_FALSE(X509Tool(args2));
+  args_list_t args3 = {"-in", in_path, "-noout", "-dates"};
+  ASSERT_FALSE(X509Tool(args3));
+  args_list_t args4 = {"-in", in_path, "-noout", "-checkend", "3600"};
```
Comment (samuel40791765) [inline tool-openssl/x509_test.cc:334]: Nit: Instead of creating a new variable for each, can we just directly reassign `args` with the new value?
```suggestion
  args = {"-in", in_path, "-noout", "-modulus"};
  ASSERT_FALSE(X509Tool(args));
  args = {"-in", in_path, "-noout", "-dates"};
  ASSERT_FALSE(X509Tool(args));
  args = {"-in", in_path, "-noout", "-checkend", "3600"};
```

File: tool-openssl/x509_test.cc [hunk 14]
```diff
@@ -47,59 +48,336 @@ X509* CreateAndSignX509Certificate() {
...
+
+  args_list_t args = {"-in", in_path, "-checkend", "3600"};
+  bool result = X509Tool(args);
+  ASSERT_TRUE(result);
+
+  RemoveFile(in_path);
+}
+
+// Test mutually exclusive options, required options, and required arugments
+TEST(X509Test, MutuallyExclusiveOptionsTest) {
```
Comment (samuel40791765) [inline tool-openssl/x509_test.cc:295]: Great test! It does seems like we can parameterize the `args` options here with `TEST_P` and C++ classes. It's just a potential improvement, but not really a priority. This might serve as a good example:

https://github.com/aws/aws-lc/blob/1fbf58437a32da4008774eed92530b787d210b1e/crypto/cipher_extra/cipher_test.cc#L591-L616

Comment (ecdeye) [reply tool-openssl/x509_test.cc:299]: Currently trying to implement but errors preventing build. Will come back to this later

**Review by justsmth — APPROVED** [review]

**Review by samuel40791765 — APPROVED** [review]

**Review by lilpoozie2005 — APPROVED** [review]
